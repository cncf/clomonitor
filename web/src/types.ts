export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
