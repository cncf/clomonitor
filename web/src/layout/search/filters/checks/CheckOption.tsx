import classNames from 'classnames';
import { isUndefined } from 'lodash';
import { GoCheck, GoX } from 'react-icons/go';

import { REPORT_OPTIONS } from '../../../../data';
import { FilterKind, ReportOption, ScoreType } from '../../../../types';
import ElementWithTooltip from '../../../common/ElementWithTooltip';
import styles from './CheckOption.module.css';

interface Props {
  type: ScoreType;
  option: ReportOption;
  activePassingChecks: ReportOption[];
  activeNotPassingChecks: ReportOption[];
  onChange: (name: FilterKind.PassingCheck | FilterKind.NotPassingCheck, value: ReportOption, checked: boolean) => void;
}

const CheckOption = (props: Props) => {
  const option = REPORT_OPTIONS[props.option];

  if (isUndefined(option)) return null;

  const isInPassingCheck = props.activePassingChecks.includes(props.option);
  const isInNotPassingCheck = props.activeNotPassingChecks.includes(props.option);

  return (
    <div className="d-flex flex-row align-items-center my-1">
      <div className={`btn-group me-2 me-sm-4 ${styles.btns}`} role="group">
        <div className="d-none d-md-flex">
          <ElementWithTooltip
            element={
              <button
                type="button"
                className={classNames('btn rounded-0 border p-0', styles.btn, styles.passingBtn, {
                  [styles.isPassing]: isInPassingCheck,
                })}
                onClick={() => {
                  props.onChange(FilterKind.PassingCheck, props.option, !isInPassingCheck);
                }}
                aria-label={`${isInPassingCheck ? 'Unselect' : 'Select'} ${option.name} as passed`}
              >
                <div className="d-flex align-items-center justify-content-center">
                  <GoCheck />
                </div>
              </button>
            }
            tooltipMessage="Check passed"
            tooltipArrowClassName={styles.iconTooltipArrow}
            alignmentTooltip="left"
            forceAlignment
            visibleTooltip
            delay={1200}
            active
            onlyInHoverableDevices
          />

          <ElementWithTooltip
            element={
              <button
                type="button"
                className={classNames('btn rounded-0 border p-0', styles.btn, {
                  [styles.isNotPassing]: isInNotPassingCheck,
                })}
                onClick={() => {
                  props.onChange(FilterKind.NotPassingCheck, props.option, !isInNotPassingCheck);
                }}
                aria-label={`${isInNotPassingCheck ? 'Unselect' : 'Select'} ${option.name} as not passed`}
              >
                <div className="d-flex align-items-center justify-content-center">
                  <GoX />
                </div>
              </button>
            }
            tooltipMessage="Check not passed"
            tooltipArrowClassName={styles.iconTooltipArrow}
            alignmentTooltip="left"
            forceAlignment
            visibleTooltip
            delay={1200}
            active
            onlyInHoverableDevices
          />
        </div>
        <div className="d-flex d-md-none">
          <button
            type="button"
            className={classNames('btn rounded-0 border p-0', styles.btn, styles.passingBtn, {
              [styles.isPassing]: isInPassingCheck,
            })}
            onClick={() => {
              props.onChange(FilterKind.PassingCheck, props.option, !isInPassingCheck);
            }}
            aria-label={`${isInPassingCheck ? 'Unselect' : 'Select'} ${option.name} as passed`}
          >
            <div className="d-flex align-items-center justify-content-center">
              <GoCheck />
            </div>
          </button>
          <button
            type="button"
            className={classNames('btn rounded-0 border p-0', styles.btn, {
              [styles.isNotPassing]: isInNotPassingCheck,
            })}
            onClick={() => {
              props.onChange(FilterKind.NotPassingCheck, props.option, !isInNotPassingCheck);
            }}
            aria-label={`${isInNotPassingCheck ? 'Unselect' : 'Select'} ${option.name} as not passed`}
          >
            <div className="d-flex align-items-center justify-content-center">
              <GoX />
            </div>
          </button>
        </div>
      </div>
      <small className="me-1 me-md-2 position-relative">{option.icon}</small>
      <div className={`me-0 me-sm-2 text-truncate flex-grow-1 ${styles.label}`}>
        <span className="d-none d-sm-block text-truncate">{option.name}</span>
        <span className="d-block d-sm-none text-truncate">{option.shortName || option.name}</span>
      </div>
    </div>
  );
};

export default CheckOption;
