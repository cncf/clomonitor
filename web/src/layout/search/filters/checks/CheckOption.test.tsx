import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ReportOption, ScoreType } from '../../../../types';
import CheckOption from './CheckOption';

const mockOnChange = jest.fn();

const defaultProps = {
  type: ScoreType.BestPractices,
  option: ReportOption.ArtifactHubBadge,
  activePassingChecks: [],
  activeNotPassingChecks: [],
  onChange: mockOnChange,
};

describe('CheckOption', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<CheckOption {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders Checks', () => {
      render(<CheckOption {...defaultProps} />);

      expect(screen.getAllByText('Artifact Hub badge')).toHaveLength(2);

      expect(screen.getAllByRole('button', { name: 'Select Artifact Hub badge as passed' })).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: 'Select Artifact Hub badge as not passed' })).toHaveLength(2);
    });

    it('renders option Artifact Hub badge as passed', () => {
      render(<CheckOption {...defaultProps} activePassingChecks={[ReportOption.ArtifactHubBadge]} />);

      expect(screen.getAllByRole('button', { name: 'Unselect Artifact Hub badge as passed' })).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: 'Select Artifact Hub badge as not passed' })).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: 'Unselect Artifact Hub badge as passed' })[0]).toHaveClass(
        'isPassing'
      );
    });

    it('renders option Artifact Hub badge as not passed', () => {
      render(<CheckOption {...defaultProps} activeNotPassingChecks={[ReportOption.ArtifactHubBadge]} />);

      expect(screen.getAllByRole('button', { name: 'Select Artifact Hub badge as passed' })).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: 'Unselect Artifact Hub badge as not passed' })).toHaveLength(2);
      expect(screen.getAllByRole('button', { name: 'Unselect Artifact Hub badge as not passed' })[0]).toHaveClass(
        'isNotPassing'
      );
    });

    it('calls onChange to click an unchecked option', async () => {
      render(<CheckOption {...defaultProps} activePassingChecks={[ReportOption.ArtifactHubBadge]} />);

      const btn = screen.getAllByRole('button', { name: 'Unselect Artifact Hub badge as passed' })[0];
      expect(btn).toBeInTheDocument();

      await userEvent.click(btn);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('passing_check', 'artifacthub_badge', false);
    });
  });
});
