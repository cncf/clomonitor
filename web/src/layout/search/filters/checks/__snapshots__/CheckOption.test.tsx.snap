// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckOption creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-row align-items-center my-1"
  >
    <div
      class="btn-group me-2 me-sm-4 btns"
      role="group"
    >
      <div
        class="d-none d-md-flex"
      >
        <div
          class="d-none d-md-block position-relative undefined"
        >
          <div
            class="cursorDefault"
            data-testid="elementWithTooltip"
          >
            <button
              aria-label="Select Artifact Hub badge as passed"
              class="btn rounded-0 border p-0 btn passingBtn"
              type="button"
            >
              <div
                class="d-flex align-items-center justify-content-center"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 12 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 5l-8 8-4-4 1.5-1.5L4 10l6.5-6.5L12 5z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          class="d-none d-md-block position-relative undefined"
        >
          <div
            class="cursorDefault"
            data-testid="elementWithTooltip"
          >
            <button
              aria-label="Select Artifact Hub badge as not passed"
              class="btn rounded-0 border p-0 btn"
              type="button"
            >
              <div
                class="d-flex align-items-center justify-content-center"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 12 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        class="d-flex d-md-none"
      >
        <button
          aria-label="Select Artifact Hub badge as passed"
          class="btn rounded-0 border p-0 btn passingBtn"
          type="button"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 12 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 5l-8 8-4-4 1.5-1.5L4 10l6.5-6.5L12 5z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Select Artifact Hub badge as not passed"
          class="btn rounded-0 border p-0 btn"
          type="button"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 12 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <small
      class="me-1 me-md-2 position-relative"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 16V8a2 2 0 0 0-1-1.73l-7-4a2 2 0 0 0-2 0l-7 4A2 2 0 0 0 3 8v8a2 2 0 0 0 1 1.73l7 4a2 2 0 0 0 2 0l7-4A2 2 0 0 0 21 16z"
        />
      </svg>
    </small>
    <div
      class="me-0 me-sm-2 text-truncate flex-grow-1 label"
    >
      <span
        class="d-none d-sm-block text-truncate"
      >
        Artifact Hub badge
      </span>
      <span
        class="d-block d-sm-none text-truncate"
      >
        Artifact Hub badge
      </span>
    </div>
  </div>
</DocumentFragment>
`;
